src/minmax.rs

```diff
@@ -0,0 +1,272 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::time::Duration;
+use std::time::Instant;
+
+#[derive(Copy, Clone)]
+struct MinmaxSample {
+    time: Instant,
+    value: Duration,
+}
+
+pub struct Minmax {
+    estimate: [MinmaxSample; 3],
+}
+
+impl Minmax {
+    pub fn new() -> Self {
+        Minmax {
+            estimate: [MinmaxSample {
+                time: Instant::now(),
+                value: Duration::new(0, 0),
+            }; 3],
+        }
+    }
+
+    /// Resets the estimates to the given value.
+    pub fn reset(&mut self, time: Instant, meas: Duration) -> Duration {
+        let val = MinmaxSample { time, value: meas };
+
+        for i in self.estimate.iter_mut() {
+            *i = val;
+        }
+        self.estimate[0].value
+    }
+
+    /// Updates the min estimate based on the given measurement, and returns it.
+    pub fn running_min(
+        &mut self, win: Duration, time: Instant, meas: Duration,
+    ) -> Duration {
+        let val = MinmaxSample { time, value: meas };
+
+        let dt = time.duration_since(self.estimate[2].time);
+
+        // Reset if there's nothing in the window or a new min value is found.
+        if val.value <= self.estimate[0].value || dt > win {
+            return self.reset(time, meas);
+        }
+
+        if val.value <= self.estimate[1].value {
+            self.estimate[2] = val;
+            self.estimate[1] = val;
+        } else if val.value <= self.estimate[2].value {
+            self.estimate[2] = val;
+        }
+
+        self.subwin_update(win, time, meas)
+    }
+
+    /// Updates the max estimate based on the given measurement, and returns it.
+    pub fn _running_max(
+        &mut self, win: Duration, time: Instant, meas: Duration,
+    ) -> Duration {
+        let val = MinmaxSample { time, value: meas };
+
+        let dt = time.duration_since(self.estimate[2].time);
+
+        // Reset if there's nothing in the window or a new max value is found.
+        if val.value >= self.estimate[0].value || dt > win {
+            return self.reset(time, meas);
+        }
+
+        if val.value >= self.estimate[1].value {
+            self.estimate[2] = val;
+            self.estimate[1] = val;
+        } else if val.value >= self.estimate[2].value {
+            self.estimate[2] = val
+        }
+
+        self.subwin_update(win, time, meas)
+    }
+
+    fn subwin_update(
+        &mut self, win: Duration, time: Instant, meas: Duration,
+    ) -> Duration {
+        let val = MinmaxSample { time, value: meas };
+
+        let dt = time.duration_since(self.estimate[0].time);
+
+        if dt > win {
```

Comment:
The original code seems to have a fair amount of comments in this function which are pretty useful to understand what it's doing, Can we keep them?

Reply:
Sure, added them.